synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -216,8 +220,42 @@ async def action_for_event_by_user(
             sender_power_level,
         ) = await self._get_power_levels_and_sender_level(event, context)
 
+        # If the experimental feature is not enabled, skip fetching relations.
+        relations = {}
+        if self._relations_match_enabled:
+            # If the event does not have a relation, then cannot have any mutual
+            # relations.
+            relation = relation_from_event(event)
+            if relation:
+                # Pre-filter to figure out which relation types are interesting.
+                rel_types = set()
+                for rule in itertools.chain(*rules_by_user.values()):
+                    # Skip disabled rules.
+                    if "enabled" in rule and not rule["enabled"]:
+                        continue
+
+                    for condition in rule["conditions"]:
+                        if condition["kind"] != "org.matrix.msc3772.relation_match":
+                            continue
+
+                        # rel_type is required.
+                        rel_type = condition.get("rel_type")
+                        if rel_type:
+                            rel_types.add(rel_type)
+
+                # If any valid rules were found, fetch the mutual relations.
+                if rel_types:
+                    relations = await self.store.get_mutual_event_relations(
+                        relation.parent_id, rel_types
+                    )
```

Comment:
Could we move this out into a helper function?

Reply:
This was done in 5a7566aab3c77f95e67d483af1a446b7642442d5.